### DISCUSSION TITLE
Support ESI

### DISCUSSION THREAD

**Discussion by unixwitch** [wagtail/wagtail#9015] [Ideas]
To improve the efficiency of front-end caching, Wagtail should support generating pages in a modular format using [edge-side includes](https://en.wikipedia.org/wiki/Edge_Side_Includes) (ESI).  This would mean global site content, such as headers and footers, could be cached (and purged) separately from individual page content.

The primary advantage of this is situations such as renaming a page which is in the site menu.  Until the cached version of every existing page expires, those pages will still show the old name in their menu, with a possibly invalid URL.  Alternatively, every page rename could purge the entire cache, but this is inefficient and significantly reduces cache lifetime, especially for large sites which rely on large caches to catch the less frequently accessed pages (where the majority of your cache hits come from the combined 'long tail' of pages which individually see few accesses).

With ESI, a typical request for a page like https://example.com/about-us would return something like this:

```
<!DOCTYPE html>
<head>
  <title>About us</title>
</head>
<body>
  <esi:include src="https://example.com/__fragments/__header">
  <div id='content'>Welcome to example.com, we provide examples.</div>
  <esi:include src="https://example.com/__fragments/__footer">
</body>
</html>
```

By issuing a cache purge for https://example.com/__fragments/__header, the header/menu could be changed for every page without invalidating the page content; the new header only needs to be re-cached once to apply to all pages.

This would also allow a different cache expiry time for the included content than individual pages, using a `Cache-Control` header.  For example, a page could be cached for a week, while the header and footer would only be cached for an hour.

Obviously, the particular modules which make up a particular page could be changed; the above is only an example to demonstrate the idea.

NB: ESI requires a front-end cache/proxy which understands the ESI directives.  We're moving toward ATS at Torchbox, which has good ESI support.  (Our previous proxy, nginx, had little or no ESI support when I last looked.)  Implementing this should allow us to significantly increase cache hit rate, which would be a huge benefit for page load times and thus UX.

cc: @balazs-endresz

**Comment by kaedroho** [comment]
I don't think theres anything Wagtail needs to do to support ESI. This should be possible already.

You'll need to create separate views/templates for each fragment in your project. Then add the `<esi:include>` tag into your page templates.

We may need to come up with some middleware to allow all the fragments to be tied together in development mode. Somebody may have done this already though.

EDIT: or use this https://addons.mozilla.org/en-us/firefox/addon/esi-processor/?src=search

**Comment by thibaudcolas** [comment]
Moving this to GitHub Discussions as this seems unlikely for us to implement in Wagtail core – we’d likely recommend implementing a package for this.
